core/src/test/scala/unit/kafka/cluster/PartitionTest.scala

```diff
@@ -230,10 +238,86 @@ class PartitionTest {
+    val controllerId = brokerId + 3
+    val replicas = List[Integer](leader, follower1, follower2).asJava
+    val isr = List[Integer](leader, follower2).asJava
+    val leaderEpoch = 8
+    val batch1 = createRecords(List(new SimpleRecord("k1".getBytes, "v1".getBytes),
+                                    new SimpleRecord("k2".getBytes, "v2".getBytes)))
+    val batch2 = createRecords(List(new SimpleRecord("k3".getBytes, "v1".getBytes),
+                                    new SimpleRecord("k4".getBytes, "v2".getBytes),
+                                    new SimpleRecord("k5".getBytes, "v3".getBytes)))
+    val batch3 = createRecords(List(new SimpleRecord("k6".getBytes, "v1".getBytes),
+                                    new SimpleRecord("k7".getBytes, "v2".getBytes)))
+
+    val partition = new Partition(topicPartition.topic, topicPartition.partition, time, replicaManager)
+    assertTrue(s"Expected first makeLeader() to return 'leader changed'",
+               partition.makeLeader(controllerId, new LeaderAndIsrRequest.PartitionState(controllerEpoch, leader, leaderEpoch, isr, 1, replicas, true), 0))
+    assertEquals(s"Current leader epoch", leaderEpoch, partition.getLeaderEpoch)
+    assertTrue(s"Expected under-replicated partitions due to one of the followers not in ISR", partition.isUnderReplicated)
+
+    // after makeLeader(() call, partition should know about all the replicas
+    val leaderReplica = partition.getReplica(leader).get
+    val follower1Replica = partition.getReplica(follower1).get
+    val follower2Replica = partition.getReplica(follower2).get
+
+    // append records with initial leader epoch
+    val lastOffsetOfFirstBatch = partition.appendRecordsToLeader(batch1, isFromClient = true).lastOffset
+    partition.appendRecordsToLeader(batch2, isFromClient = true)
+    assertEquals(s"Expected leader's HW not move", leaderReplica.logStartOffset, leaderReplica.highWatermark.messageOffset)
+
+    // let the follower in ISR move leader's HW to move further but below LEO
+    def readResult(fetchInfo: FetchDataInfo, leaderReplica: Replica): LogReadResult = {
+      LogReadResult(info = fetchInfo,
+                    highWatermark = leaderReplica.highWatermark.messageOffset,
+                    leaderLogStartOffset = leaderReplica.logStartOffset,
+                    leaderLogEndOffset = leaderReplica.logEndOffset.messageOffset,
+                    followerLogStartOffset = 0,
+                    fetchTimeMs = time.milliseconds,
+                    readSize = 10240,
+                    lastStableOffset = None)
+    }
+    partition.updateReplicaLogReadResult(
+      follower2Replica, readResult(FetchDataInfo(LogOffsetMetadata(0), batch1), leaderReplica))
+    partition.updateReplicaLogReadResult(
+      follower2Replica, readResult(FetchDataInfo(LogOffsetMetadata(lastOffsetOfFirstBatch), batch2), leaderReplica))
+    assertEquals(s"Expected leader's HW", lastOffsetOfFirstBatch, leaderReplica.highWatermark.messageOffset)
```

Comment:
nit: there are a few assertions without any substitutions in the message. We can drop the `s` prefix.